File: demo/src/main/java/com/example/demo/service/UserService.java

package com.example.demo.service;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import com.example.demo.dto.request.UserRequest;
import com.example.demo.dto.response.UserResponse;
import com.example.demo.entity.User;
import com.example.demo.enums.AccountStatus;
import com.example.demo.enums.Role;
import com.example.demo.exception.AppException;
import com.example.demo.exception.ErrorCode;
import com.example.demo.exception.ResourceNotFoundException;
import com.example.demo.mapper.UserMapper;
import com.example.demo.repository.UserRepository;
import com.example.demo.security.UserPrincipal;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
@Slf4j
public class UserService {
    UserRepository userRepository;
    PasswordEncoder passwordEncoder;
    MailService mailService;
    ImageStorageService imageStorageService;

    public UserResponse userRegister(UserRequest userRqDto) {
        User user = UserMapper.toEntity(userRqDto);
        if (userRepository.findByEmail(userRqDto.getEmail()).isPresent()) {
            throw new AppException(ErrorCode.USER_EXISTED);
        }
        if (userRepository.findByUserName(userRqDto.getUserName()).isPresent()) {
            throw new AppException(ErrorCode.USER_EXISTED);
        }
        user.setPassword(passwordEncoder.encode(userRqDto.getPassword()));

        userRepository.save(user);
        return UserMapper.toResponse(user);
    }

    public UserResponse getUserInfo(UserRequest userRqDto) {
        return userRepository.findByUserName(userRqDto.getUserName()).map(UserMapper::toResponse)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
    }

    public UserResponse getUserById(String id) {
        return userRepository.findById(id).map(UserMapper::toResponse)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
    }

    public UserDetails loadUserById(String id) {
        User user = userRepository.findById(id).orElseThrow(
                () -> new ResourceNotFoundException("User", "id", id));

        return UserPrincipal.create(user);
    }

    public UserResponse uploadAvatar(String userId, MultipartFile file) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        String path = imageStorageService.storeAvatar(userId, file);

        user.setImageUrl(path);
        userRepository.save(user);
        return UserMapper.toResponse(user);
    }

    public UserResponse updateUserProfile(String userId, UserRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        if (request.getName() != null) {
            user.setName(request.getName());
        }
        if (request.getPhoneNumber() != null) {
            user.setPhoneNumber(request.getPhoneNumber());
        }
        if (request.getCccd() != null) {
            user.setCccd(request.getCccd());
        }

        userRepository.save(user);
        log.info("User profile updated for ID: {}", userId);
        return UserMapper.toResponse(user);
    }

    public UserResponse changePassword(String userId, String oldPassword, String newPassword) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        if (!passwordEncoder.matches(oldPassword, user.getPassword())) {
            throw new AppException(ErrorCode.USER_EXISTED);
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);

        log.info("Password changed for user: {}", userId);
        return UserMapper.toResponse(user);
    }

    public List<UserResponse> getAllUsers() {
        return userRepository.findAll().stream()
                .map(UserMapper::toResponse)
                .collect(Collectors.toList());
    }

    public UserResponse getUserByEmail(String email) {
        return userRepository.findByEmail(email)
                .map(UserMapper::toResponse)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));
    }

    public UserResponse updateAccountStatus(String userId, AccountStatus status) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        user.setStatus(status);
        userRepository.save(user);

        log.info("Account status updated to {} for user: {}", status, userId);
        return UserMapper.toResponse(user);
    }

    public UserResponse updateUserRole(String userId, Role role) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException(ErrorCode.USER_NOT_FOUND));

        user.setRole(role);
        userRepository.save(user);

        log.info("Role updated to {} for user: {}", role, userId);
        return UserMapper.toResponse(user);
    }

    public void deleteUser(String userId) {
        if (!userRepository.existsById(userId)) {
            throw new AppException(ErrorCode.USER_NOT_FOUND);
        }

        userRepository.deleteById(userId);
        log.info("User deleted: {}", userId);
    }

}

File: fe/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import {
  ApiResponse,
  UserResponse,
  UserRequest,
  AccountStatus,
  Role,
} from '../models/api-response.model';

@Injectable({
  providedIn: 'root',
})
export class UserService extends ApiService {
  getUserById(id: string): Observable<ApiResponse<UserResponse>> {
    return this.get<ApiResponse<UserResponse>>(`/user/${id}`);
  }

  getUserByEmail(email: string): Observable<ApiResponse<UserResponse>> {
    return this.get<ApiResponse<UserResponse>>(`/user/email/${email}`);
  }

  getAllUsers(): Observable<ApiResponse<UserResponse[]>> {
    return this.get<ApiResponse<UserResponse[]>>('/user');
  }

  updateUserProfile(id: string, request: UserRequest): Observable<ApiResponse<UserResponse>> {
    return this.put<ApiResponse<UserResponse>>(`/user/${id}`, request);
  }

  changePassword(
    id: string,
    oldPassword: string,
    newPassword: string
  ): Observable<ApiResponse<UserResponse>> {
    return this.put<ApiResponse<UserResponse>>(
      `/user/${id}/password?oldPassword=${oldPassword}&newPassword=${newPassword}`,
      {}
    );
  }

  uploadAvatar(id: string, file: File): Observable<ApiResponse<UserResponse>> {
    const formData = new FormData();
    formData.append('file', file);
    return this.uploadFile<ApiResponse<UserResponse>>(`/user/${id}/avatar`, formData);
  }

  updateAccountStatus(id: string, status: AccountStatus): Observable<ApiResponse<UserResponse>> {
    return this.put<ApiResponse<UserResponse>>(`/user/${id}/status?status=${status}`, {});
  }

  updateUserRole(id: string, role: Role): Observable<ApiResponse<UserResponse>> {
    return this.put<ApiResponse<UserResponse>>(`/user/${id}/role?role=${role}`, {});
  }

  deleteUser(id: string): Observable<ApiResponse<string>> {
    return this.delete<ApiResponse<string>>(`/user/${id}`);
  }
}
